[Serialization] Recover from ObjC protocols changing inheritance.

jrose-apple opened the pull request:
When there's an Objective-C protocol that adopts other protocols, the other protocols become part of the requirement signature. If that can change, Swift conformances to that protocol will get very confused when it comes time to deserialize the conformances that satisfy the requirement signature.

To recover from this, just deserialize *all* trailing conformances, rather than follow the requirement signature, and match them up after the fact. (This only works for Objective-C protocols where we know all conformance requirements represent inherited protocols, as opposed to constraints on associated types.)

rdar://problem/33356098

jrose-apple commented:
@swift-ci Please test

DougGregor commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -4591,17 +4591,57 @@ void ModuleFile::finishNormalConformance(NormalProtocolConformance *conformance,
+      } else {
+        // Put in an abstract conformance as a placeholder. This is a lie, but
+        // there's not much better we can do.
+        reqConformances.push_back(ProtocolConformanceRef(proto));
```
This is likely to blow up further down the line, because we won't expect an abstract conformance here. Can we do a lookup to find the conformance we need?

jrose-apple replied:
In my test case it exists, but there's no guarantee that it actually does. (And I should add a test case where it doesn't.) I'm also concerned that that might find a conformance in *another* module as well. The other idea I had discussed with @jckarter was to make up a completely opaque conformance, since it is an Objective-C protocol.

(This is not at all a safe change to make even in Objective-C. I just don't want the compiler to crash.)

DougGregor replied:
Plumbing through a completely opaque conformance for this case seems... worse. Thinking more about the abstract conformance, we *might* be okay in practice, because (1) the substitution machinery won't ever need to look at this conformance due to the lack of associated types in `@objc` protocols, and (2) IRGen never needs to fulfill a conformance to an `@objc` protocol, so it doesn't care that there's an abstract conformance.

jrose-apple replied:
I can't say I'm happy about this, but right now we're just aiming for "not crash".

jckarter replied:
(2) was my thinking. We can probably skate changes like this by as long as we don't look too closely at ObjC protocol hierarchies, and for the most part we don't usually have to.

DougGregor commented on lib/Serialization/Deserialization.cpp:
```diff
@@ -4591,17 +4591,57 @@ void ModuleFile::finishNormalConformance(NormalProtocolConformance *conformance,
-      reqConformances.push_back(reqConformance);
+
+  if (proto->isObjC() && getContext().LangOpts.EnableDeserializationRecovery) {
+    // Be resilient against inherited protocols being added or removed.
```
Adding or removing an inherited protocol is going to cause mayhem with generic signatures involving the inheriting protocol. It's not at all a resilient change. That said, this is an improvement for a very specific problem, and we can consider the larger issue as part of resilience.

jrose-apple replied:
Yeah, maybe I should avoid the word "resilient".

DougGregor replied:
s/be resilient/Don't crash when.

jrose-apple committed:
[Serialization] Recover from ObjC protocols changing inheritance.

When there's an Objective-C protocol that adopts other protocols, the
other protocols become part of the requirement signature. If that can
change, Swift conformances to that protocol will get very confused
when it comes time to deserialize the conformances that satisfy the
requirement signature.

To recover from this, just deserialize /all/ trailing conformances,
rather than follow the requirement signature, and match them up after
the fact. (This only works for Objective-C protocols where we know all
conformance requirements represent inherited protocols, as opposed to
constraints on associated types.)

rdar://problem/33356098

jrose-apple commented:
Tweaked the comments, beefed up the test some, fixed the merge conflict.

@swift-ci Please smoke test
